### DISCUSSION TITLE
arthas-agent模块的bindingThread的设计具体是为了解决什么场景下的问题？

### DISCUSSION THREAD

**Discussion by GayITHub** [alibaba/arthas#2474] [Q&A]
注释上写的是： Use a dedicated thread to run the binding logic to prevent possible memory leak. 

jvm在加载agent的时候 本身就是一个单独的线程来执行，按理来说agent加载完成之后，加载线程就会被销毁并且释放相关内存资源。
这里给bind动作又开了一个单独的线程，不是很能想明白什么场景下才会出现注释中说的可能的内存泄漏

**Comment by hengyunabc** [comment]
下面是当时的记录

---

#### 泄漏点1:

netty的初始化逻辑放在agent attach的线程`Attach Listener`（JVM级别的线程）中，netty初始化的时候会创建自定义的实现ThreadLocal， 因此`io.netty.util.internal.InternalThreadLocalMap`这个类会`Attach Listener`线程引用，而无法回收

#### 解决办法

1. 不要在`Attach Listener`线程中进行netty的初始化，放在异步的线程中
2. netty关闭的时候，清除`Attach Listener`的引用，但这个很难做到，我们没有机会在Attach Listener中再执行代码了。

![image](https://user-images.githubusercontent.com/1683936/235099288-6c45ae80-1ea1-4fba-a224-63c0d715e5b1.png)

#### 泄漏点2

清理session的线程在Arthas 停止的过程中没有shutdown。
